src/Features/Core/Portable/SolutionCrawler/SolutionCrawlerProgressReporter.cs

```diff
@@ -19,109 +16,69 @@ internal partial class SolutionCrawlerRegistrationService : ISolutionCrawlerRegi
+            private void Pause() => ChangeProgressStatus(ref _progressEvaluateCount, ProgressStatus.Paused);
 
-                _count = 0;
-            }
 
-            public bool InProgress
+            public void UpdatePendingItemCount(int pendingItemCount)
             {
-                get
+                if (_progressStartCount > 0)
                 {
-                    return _count > 0;
+                    var progressData = new ProgressData(ProgressStatus.PendingItemUpdated, pendingItemCount);
+                    OnProgressChanged(progressData);
                 }
             }
 
-            public event EventHandler<ProgressData> ProgressChanged
+            public IDisposable Evaluating()
             {
-                add
-                {
-                    _eventMap.AddEventHandler(nameof(ProgressChanged), value);
-                }
-
-                remove
-                {
-                    _eventMap.RemoveEventHandler(nameof(ProgressChanged), value);
-                }
+                return new ProgressStatusRAII(this);
             }
 
-            public Task Start()
+            private void ChangeProgressStatus(ref int referenceCount, ProgressStatus status)
             {
-                if (Interlocked.Increment(ref _count) == 1)
+                var start = (status == ProgressStatus.Started || status == ProgressStatus.Evaluating);
+                if (start ? (Interlocked.Increment(ref referenceCount) == 1) : (Interlocked.Decrement(ref referenceCount) == 0))
                 {
-                    _lastReportedFilePath = null;
-
-                    var asyncToken = _listener.BeginAsyncOperation("ProgressReportStart");
-                    var progressData = new ProgressData(ProgressStatus.Started, filePathOpt: null);
-                    return RaiseEvent(nameof(ProgressChanged), progressData).CompletesAsyncOperation(asyncToken);
+                    var progressData = new ProgressData(status, pendingItemCount: null);
+                    OnProgressChanged(progressData);
                 }
-
-                return Task.CompletedTask;
             }
 
-            public Task Stop()
+            private void OnProgressChanged(ProgressData progressData)
             {
-                if (Interlocked.Decrement(ref _count) == 0)
-                {
-                    _lastReportedFilePath = null;
-
-                    var asyncToken = _listener.BeginAsyncOperation("ProgressReportStop");
-                    var progressData = new ProgressData(ProgressStatus.Stoped, filePathOpt: null);
-                    return RaiseEvent(nameof(ProgressChanged), progressData).CompletesAsyncOperation(asyncToken);
-                }
-
-                return Task.CompletedTask;
+                ProgressChanged?.Invoke(this, progressData);
             }
 
-            public Task Update(string filePath)
+            private struct ProgressStatusRAII : IDisposable
```

Comment:
ProgressStatusResumer?   Shoudl have documentation on what it's semantics actually are.  It was somewhat surprising to me that it caused Evaluate/Pause to be called.

Reply:
I dont think ProgressStatusResumer make it more clear. name doesn't matter mucn if we had no name type that I can temporary can create, I would have used that. all I care is it is RAII so that I don't need to care calling counter part method on exception.

Reply:
Sounds good.  Thanks.